tests/foreman/cli/test_fact.py

```diff
@@ -100,16 +99,27 @@ def test_positive_update_client_facts_verify_imported_values(
         activation_keys=[module_activation_key.name],
     )
     assert result.status == 0, f'Failed to register host: {result.stderr}'
-    rhel_contenthost.execute('subscription-manager facts --update')
+
+    host = rhel_contenthost.nailgun_host
+    # gather ansible facts by running ansible roles on the host
+    task_id = host.play_ansible_roles()
+    module_target_sat.wait_for_tasks(
+        search_query=f'id = {task_id}',
+        poll_timeout=100,
+    )
+    task_details = module_target_sat.api.ForemanTask().search(query={'search': f'id = {task_id}'})
+    assert task_details[0].result == 'success'
+    assert rhel_contenthost.execute('subscription-manager facts --update').status == 0
```

Comment:
I see that the facts update was there before, but can you please elaborate on why the facts update is needed?

Reply:
I have added ansible facts too so updating the facts to Satellite after gathering all the facts. Also, `play_ansible_roles()` doesn't verify if the job is complete so the facts were updating before the job is complete. So I added a check to verify the success of ansible run before uploading the facts to Satellite

Reply:
I do not see how `subscription-manager facts --update` is related to the fact that you played Ansible roles which made the [foreman_callback](https://docs.ansible.com/ansible/latest/collections/theforeman/foreman/foreman_callback.html) plugin report facts, gathered by Ansible, to report the facts to Satellite.

As far as I know, there are three main sources of facts in Satellite.
1. Ansible
2. subscription-manager
3. Puppet

Therefore I would like to know what should be the effect of running `subscription-manager facts --update`. Shouldn't system facts be uploaded to Satellite by the subscription-manager upon registration? What has changed on the system that we have to update subscription-manager facts?

Reply:
Facts get uploaded on various points like when we do host registration, `subscription-manager facts --update` is run as part of script. But we have an issue that the facts does not get uploaded on registration unless explicitly uploaded or it is uploaded to Satellite every 4 hours.

Reply:
Also, it's not just ansible facts gathered but other host facts gathered will be uploaded to Satellite. So `subscription-manager facts --update` is not run for Ansible facts .

Reply:
Hello,

subscription-manager developer here o/

> But we have an issue that the facts does not get uploaded on registration unless explicitly uploaded or it is uploaded to Satellite every 4 hours.

I'm puzzled by this sentence, would you please explain it a bit more?

To clarify the situation a bit:

1. subscription-manager collects the facts on the system from various sources, and adding the facts in the `/etc/rhsm/facts/*.facts` files
2. subscription-manager sends its whole set of facts as part of the registration
  - because of this, I don't think the sentence above matches the reality
3. when subscription-manager performs its regular check-in (by default every 4h with some random delay), it collects the facts again (see (1) above); then it checks whether the newly collected facts are the same as the last set of facts collected: _only_ in case any of the fact changed the new set of facts is sent as part of the check-in, replacing what is stored in Candlepin
4. `subscription-manager facts --update` forces a new collection of the facts (again, see (1) above), and sends them to Candlepin to replace what is the set of facts there; no comparison is done in this case

This means that:
- after `subscription-manager register` completes successfully, the rhsm facts locally and in Candlepin are in-sync
- unless anything that changes the subscription-manager facts (e.g. network interfaces, system version, or even `.facts` files), running `subscription-manager facts --update` is redundant
- unless the Ansible facts are `.facts` files for subscription-manager, they won't change the facts of subscription-manager (see the point above), and thus `subscription-manager facts --update` is not needed

Hope it clarifies a bit the situation -- I'm more than happy to clarify/explain things more as needed/wanted.

Reply:
Thanks @ptoscano for detailed explanation. I had the same understanding for facts collection and upload by looking at the code for registration. The facts collected  in `/etc/rhsm/facts/*.facts` are uploaded but the behavior was not consistent so I am explicitly uploading the facts.

Reply:
The fact upload works now but I saw inconsistent behavior in the past so I have removed it for now and let's wait for the test to break again :)

Reply:
In such cases, please report to us (client tools team) before adding workarounds in layered products that use subscription-manager. Otherwise
- we will never know there was an issue
- even in case a workaround is needed, we can provide a better/more fitting workaround